src/libs/vehicle/mavlink/vehicle.ts

```diff
@@ -773,6 +774,68 @@ export abstract class MAVLinkVehicle<Modes> extends Vehicle.AbstractVehicle<Mode
     sendMavlinkMessage(heartbeatMessage)
   }
 
+  /**
+   * Send system time from GCS
+   */
+  sendSystemTime(): void {
+    const systemTimeMessage: Message.SystemTime = {
+      type: MAVLinkType.SYSTEM_TIME,
+      time_unix_usec: Math.floor(Date.now() * 1000),
+      time_boot_ms: Math.floor(Date.now() / 1000),
+    }
+    sendMavlinkMessage(systemTimeMessage)
+  }
+
+  /**
+   * Measure network latency to vehicle using TIMESYNC message
+   * @returns {Promise<void>}
+   */
+  async measureLatency(): Promise<void> {
+    const time_sent = Math.floor(Date.now()) * 1000
+    let receivedTimesync = false
+    let timeoutReached = false
+    let latencyUs = 0
+    const timeout = 10000
+
+    const timesyncHandler = (pack: Package): void => {
+      const timesync = pack.message as Message.Timesync
+      if (timesync.ts1 === time_sent) {
+        const time_received = Math.floor(Date.now()) * 1000
```

Comment:
Should this potentially use [`Performance.now()`](https://developer.mozilla.org/en-US/docs/Web/API/Performance/now) instead, which has sub-millisecond precision, and is not subject to clock adjustments?

That would also potentially allow it to be separated from the `time_sent` variable, with the starting point instead defined just before (or after?) sending the `TIMESYNC` message. I'm unsure what's more helpful/realistic out of that vs including the time to package and allocate a handler for the message in the latency time...